web-crawler/src/main/java/com/google/sps/webcrawler/WebCrawler.java

```
@@ -0,0 +1,231 @@
+   *    "http/examples/client/ClientWithResponseHandler.java">Code reference</a>
+   */
+  public List<String> getUrlsFromCustomSearch(String candidateName) {
+    List<String> urls;
+    String request =
+        String.format("https://www.googleapis.com/customsearch/v1?key=%s&cx=%s&q=%s",
+            CUSTOM_SEARCH_KEY, CUSTOM_SEARCH_ENGINE_ID, candidateName.replace(" ", "%20"));
+    CloseableHttpClient httpclient = HttpClients.createDefault();
+    try {
+      HttpGet httpGet = new HttpGet(request);
+      ResponseHandler<String> responseHandler = new ResponseHandler<String>() {
+        @Override
+        public String handleResponse(final HttpResponse response) throws IOException {
+          int status = response.getStatusLine().getStatusCode();
+          if (status >= 200 && status < 300) {
+              HttpEntity entity = response.getEntity();
+              return entity != null ? EntityUtils.toString(entity) : null;
+          } else {
+              throw new ClientProtocolException("Unexpected response status: " + status);
+          }
+        }
+      };
+      String responseBody = httpclient.execute(httpGet, responseHandler);
+      httpclient.close();
+      Gson gson = new Gson();
+      Object jsonResponse = gson.fromJson(responseBody, Object.class);
+      //@TODO [Unpack {@code jsonResponse} and find URLs.]
+      return Arrays.asList(
+          "https://www.cnn.com/2020/06/23/politics/aoc-ny-primary-14th-district/index.html");
+    } catch (IOException e){
+      System.out.println("[ERROR] Error occurred with fetching URLs from Custom Search: " + e);
+      return Arrays.asList(
+          "https://www.cnn.com/2020/06/23/politics/aoc-ny-primary-14th-district/index.html");
+    }
+  }
+
+  /** 
+   * Checks robots.txt for permission to web-scrape, scrapes webpage if permitted and extracts
+   * textual content.
+   */
+  public NewsArticle scrapeAndExtractHtml(String url) {
```

Probably better to return an `Optional<NewsArticle>` instead of returning empty news articles upon failure:
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html